Allow in graph einops operators

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99582
* __->__ #99478

Coordinating with @arogozhnikov from einops team, allowing specific operators in the dynamo graph avoids dynamo tracing problems provided the operators are screened for safety - they must not bake in unintended constants or take data-dependent control flow paths.

Fixes #99031

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

wconstab committed:
Allow in graph einops operators

Coordinating with @arogozhnikov from einops team, allowing specific operators in the dynamo graph avoids dynamo tracing problems provided the operators are screened for safety - they must not bake in unintended constants or take data-dependent control flow paths.

Fixes #99031

[ghstack-poisoned]

jansel commented on torch/_dynamo/__init__.py:
```diff
@@ -205,3 +205,21 @@ def mark_static(t, index=None):
+
+
+def allow_in_graph_einops():
+    try:
+        from einops.einops import einsum, rearrange, reduce, repeat
+        from einops.packing import pack, unpack
+
+        allow_in_graph(rearrange)
+        allow_in_graph(reduce)
+        allow_in_graph(repeat)
+        allow_in_graph(einsum)
+        allow_in_graph(pack)
+        allow_in_graph(unpack)
+    except ImportError:
+        pass
```
From https://github.com/arogozhnikov/einops/issues/250#issuecomment-1514161763
```suggestion
    try:
        import einops
        try:
            from einops._torch_specific import allow_ops_in_compiled_graph  # requires einops>=0.6.1, torch >= 2.0
            # einops >= 0.6.1
            allow_ops_in_compiled_graph()
        except ImportError:
            # einops < 0.6.1
            allow_in_graph(einops.rearrange)
            allow_in_graph(einops.reduce)
            if hasattr(einops, 'repeat'): 
                allow_in_graph(einops.repeat) # available since einops 0.2.0 
            if hasattr(einops, 'einsum'):
                allow_in_graph(einops.einsum) # available since einops 0.5.0
            if hasattr(einops, 'pack'):
                allow_in_graph(einops.pack)   # available since einops 0.6.0
            if hasattr(einops, 'unpack'):
                allow_in_graph(einops.unpack) # available since einops 0.6.0
    except ImportError:
        pass
```

This will allow the einops folks to change this behavior in future versions.

jansel commented on torch/_dynamo/__init__.py:
```diff
@@ -205,3 +205,21 @@ def mark_static(t, index=None):
+        pass
+
+
+allow_in_graph_einops()
```
Let's move this to allowed_functions.py

jansel commented on torch/_dynamo/skipfiles.py:
```diff
@@ -171,8 +171,6 @@ def check(filename, allow_torch=False):
 # skip common third party libs
 for _name in (
-    "einops",
-    "einops_exts",
```
Can you check if `einops_exts` is still needed for:
```
./benchmarks/dynamo/torchbench.py --performance --inference -k DALLE2_pytorch
```

jansel replied:
I suspect https://github.com/pytorch/pytorch/pull/99207 will actually fix that issue, so maybe we can remove `einops_exts` and land the `ConstantVariable` part of #99207.

wconstab replied:
sure. i guess up to you guys, if we land this first i can do a follow up once 99207 lands. or i can just wait on that and rebase this

jansel replied:
We can land this.  `DALLE2_pytorch` isn't working in inductor anyway, so it won't move the stats.

arogozhnikov replied:
> is it useful for me to land this PR and keep einops_ext skipped?

@wconstab  yes, it is useful 

`einops_ext` is a separate module with additional functions by @lucidrains , so maybe he can speak to the best way forward
(I am not sure @-tagging , I can ping him via email)

wconstab replied:
ok, well, we ended up leaving einops_ext un-skipped and hopefully will be landing a fix later that addresses string-formatting issues that broke to_/from_ ops there.  If we come up with additional issues later we can address them then.

lucidrains replied:
@arogozhnikov hey yes! that package predates your addition of `pack` and `unpack`

i've removed it from both dalle2-pytorch and imagen-pytorch this morning

do let me know if you find any more traces of it in other packages

wconstab committed:
Update on "Allow in graph einops operators"


Coordinating with arogozhnikov from einops team, allowing specific operators in the dynamo graph avoids dynamo tracing problems provided the operators are screened for safety - they must not bake in unintended constants or take data-dependent control flow paths.

Fixes #99031

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Allow in graph einops operators"


Coordinating with arogozhnikov from einops team, allowing specific operators in the dynamo graph avoids dynamo tracing problems provided the operators are screened for safety - they must not bake in unintended constants or take data-dependent control flow paths.

Fixes #99031

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Allow in graph einops operators"


Coordinating with arogozhnikov from einops team, allowing specific operators in the dynamo graph avoids dynamo tracing problems provided the operators are screened for safety - they must not bake in unintended constants or take data-dependent control flow paths.

Fixes #99031

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Allow in graph einops operators"


Coordinating with arogozhnikov from einops team, allowing specific operators in the dynamo graph avoids dynamo tracing problems provided the operators are screened for safety - they must not bake in unintended constants or take data-dependent control flow paths.

Fixes #99031

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

kit1980 commented:
revert -m "dynamo/test_after_aot.py::TestAfterAot::test_save_graph_repro - AssertionError, see https://github.com/pytorch/pytorch/actions/runs/4750274195/jobs/8438535867" -c landrace

kit1980 commented:
@pytorchbot revert -m "dynamo/test_after_aot.py::TestAfterAot::test_save_graph_repro - AssertionError, see https://github.com/pytorch/pytorch/actions/runs/4750274195/jobs/8438535867" -c landrace

wconstab committed:
Update on "Allow in graph einops operators"


Coordinating with arogozhnikov from einops team, allowing specific operators in the dynamo graph avoids dynamo tracing problems provided the operators are screened for safety - they must not bake in unintended constants or take data-dependent control flow paths.

Fixes #99031

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab commented:
I will be relanding this, have to wait for https://github.com/pytorch/pytorch/pull/99627 to land first.

Note: i couldn't repro locally any way that my einops changes cause cuda context to be initialized.  I can pass @ezyang's test that asserts no cuda context initialized.

Best guess is some combination of tests being run in CI gets reordered or something and causes cuda context to be initialized when not expected.   Going to ignore and land once the assert is out.

wconstab commented:
Actually @lucidrains I want to propose a slight change to this. It would be nice if eventually, we could delete the whole block of code on our side.  (Once we figure enough time has passed that most users have picked up recent einops versions).

To accomplish this, I wonder if you could change your side to always call `allow_ops_in_compiled_graph` so that if our users import einops > 0.6.1 then we don't have to do anything special.

I'd still land this block of code that imports einops and checks for allow_ops_in_compiled_graph  and if it exists, do nothing. else, run through and allow specific functions in graph.  I'd also add a comment that we should delete this block eventually and prefer vendors call the api on their side.

Maybe we need to change the name of `allow_ops_in_compiled_graph` if you have to bump version again to accomplish this.

lucidrains commented:
@wconstab yea sure, just updated dalle2 and imagen for starters, but can start doing this going forward for some of my newer works

wconstab commented:
https://github.com/pytorch/pytorch/pull/99631
new PR @lucidrains -- lmk if I should update it to import a different function name or not.

lucidrains commented:
@wconstab yea that looks good

i'll enforce einops 0.6.1 going forward and invoke `_allow_in_graph_einops`

as for all the existing repos, i'll upgrade them on a case by case basis
